salt/fileclient.py

```diff
@@ -1207,11 +1213,12 @@ def get_file(self,
                             saltenv,
                             cachedir=cachedir) as cache_dest:
                         dest = cache_dest
+                        dest_tmp = "{0}/{1}".format(os.path.dirname(dest), str(uuid.uuid4()))
                         # If a directory was formerly cached at this path, then
                         # remove it to avoid a traceback trying to write the file
                         if os.path.isdir(dest):
                             salt.utils.files.rm_rf(dest)
-                        fn_ = salt.utils.files.fopen(dest, 'wb+')
+                        fn_ = salt.utils.files.fopen(dest_tmp, 'wb+')
```

Comment:
Instead of using `dest_tmp` here and renaming manually, you should use `salt.utils.atomicfile.atomic_open()`, which was designed exactly for this sort of use case. You can change this line to:

```python
fn_ = salt.utils.atomicfile.atomic_open(dest, 'wb+')
```

Then, when the filehandle is closed, the tempfile will be automatically renamed to the `dest` path.

Note that using this method will require two additional changes:

1. `salt.utils.atomicfile` will need to be imported here in fileclient.py.
2. `salt/utils/atomicfile.py` will need to be added to salt/utils/thin.py in the `min_files` tuple so that it is part of the thin tarball.

Reply:
Thanks for your comments @terminalmage , please see changes in the latest commit.